### PR TITLE
feat: Skip bot-comment when onboarding is skipped

### LINKED ISSUE #7281: [Bug]Skip bot comments if the comments tour is skipped
## Description
Skip bot comments if the comments tour is skipped

### Steps to reproduce the behaviour:

1. Signup with a new user
2. Click on toggle comment mode button
3. Click on the skip button, enter the name and submit the comments onboarding form
4. Add a new comment, a bot comment is added, if the tour is skipped the bot comment should also be skipped

### PR DESCRIPTION
**Pull request by nayan-rafiq**
## Description
Skips the bot comments if user has skipped the on-boarding flow during their first comment.

Fixes #7281

## Type of change
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
- Tested on local machine
- Unit tests

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/UserController.java [hunk 1]
```diff
@@ -206,4 +208,20 @@ public UserController(UserService service,
...
+    @PatchMapping("/comment/state")
+    public Mono<ResponseDTO<CommentState>> setCommentState(@RequestBody UserData userData) {
+        return userDataService.setCommentState(userData.getCommentState())
+                .map(savedUserData ->
+                        new ResponseDTO<>(HttpStatus.OK.value(), savedUserData.getCommentState(), null)
+                );
+    }
+
+    @GetMapping("/comment/state")
+    public Mono<ResponseDTO<CommentState>> getCommentState() {
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/UserController.java:220]: Is the client going to call this API on every page load? Can we put this information as part of user profile itself? That is anyway being fetched, so we save on an extra API call no?

Also, if we do that, then the client will need to make sure this field is **not** passed on to `appsmith.user` object, or folks will be able to see this in auto-complete when writing JS in Appsmith.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/constants/CommentState.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.appsmith.server.constants;
+
+public enum CommentState {
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/constants/CommentState.java:3]: If this is going to be _only_ about onboarding, can we name it `CommentOnboardingState`? Right now this name feels like the _state of a comment_, with options like `RESOLVED` meaning that the comment is _in a resolved state_, etc.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/UserController.java (hunk 1 above)
Comment (nayan-rafiq) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/UserController.java:220]: @sharat87 I thought about that but currently, we don't return any properties of `UserData` to any API e.g. `/me`. If I return `userData` contents as part of the `/me` API, other properties of the `UserData` will be exposed and many of them are for internal use. That's why I came up with a new API. After reading your comment, I think I can expose this property only in `me` API. Do you think that'll be a better idea? 

To avoid an API call each time, client will cache this value and only make an API call when cache is empty. Client will use this value to decide whether to show the on-boarding flow to user or not.

Comment (nayan-rafiq) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/UserController.java]: @sharat87 I've updated the PR. Now the `commentOnboardingState` is returned in the response of `/me` API.

**Review by sharat87 — APPROVED** [review]
Looks good, so approving. However, note that once this goes in, the `appsmith.user` object will have a `commentOnboardingState` field for all users to see and use. The client code needs to not expose that field to user-written Javascript. That change can go along with this PR togather, or soon after, but needs to go soon.

**Review by sharat87 — APPROVED** [review]

File: app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/UserController.java [hunk 3]
```diff
@@ -144,7 +146,8 @@ public UserController(UserService service,
     @GetMapping("/me")
     public Mono<ResponseDTO<UserProfileDTO>> getUserProfile() {
         return sessionUserService.getCurrentUser()
-                .flatMap(service::buildUserProfileDTO)
+                .zipWith(userDataService.getForCurrentUser())
```
Comment (mohanarpit) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/UserController.java:149]: Any particular reason that we are doing this logic in the controller & not the service method itself?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/UserController.java [hunk 4]
```diff
@@ -206,4 +209,12 @@ public UserController(UserService service,
                 .map(map -> new ResponseDTO<>(HttpStatus.OK.value(), map, null));
     }
 
+    @PatchMapping("/comment/state")
+    public Mono<ResponseDTO<CommentOnboardingState>> setCommentState(@RequestBody UserData userData) {
+        return userDataService.setCommentState(userData.getCommentOnboardingState())
```
Comment (mohanarpit) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/UserController.java:214]: Quick question: Do these DB calls not require any ACL? Just checking.

Comment (nayan-rafiq) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/UserController.java:214]: We don't set policies to `UserData` and there is no ACL check anywhere for `UserData`. Should we change that?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/UserController.java (hunk 3 above)
Comment (mohanarpit) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/UserController.java:149]: Yes please. If we need to refactor, lets do that. Adding such logic in a controller makes it virtually untestable since we don't have any controller level tests in the code base.

**Review by sharat87 — CHANGES_REQUESTED** [review]
Looks good. It looks like all tests have a `@WithUserDetails(value = "api_user")` above them. Considering the error we saw recently, can we add a test where no one is logged in, and ensure that we get a profile object, instead of an empty Mono?

**Comment by nayan-rafiq** [comment]
> Looks good. It looks like all tests have a `@WithUserDetails(value = "api_user")` above them. Considering the error we saw recently, can we add a test where no one is logged in, and ensure that we get a profile object, instead of an empty Mono?

@sharat87 I've added a unit test to test profile is returned when user is anonymous.

**Review by sharat87 — APPROVED** [review]
Looks good to be merged. Please:

1. Edit the merge commit's detail to the PR description, or otherwise more details about the change.
1. Squash merge once the CI is finished. Thanks!
